scripts/dats_jsonld_annotator/annotator.py

```diff
@@ -0,0 +1,323 @@
+            _schema = find_schema(schema, k, v)
+            json_object[k], _local_context = annotate_dats_object(
+                v, _schema, specific_context, context_dir
+            )
+            specific_context.update(_local_context)
+        if isinstance(v, list):
+            # Let's find the schema link under "items"
+            annotation_list = []
+            for vv in v:
+                if not isinstance(vv, dict):
+                    annotation_list.append(vv)
+                    continue
+                _schema = find_schema(schema, k, vv)
+                _json_o, _local_context = annotate_dats_object(
+                    vv, _schema, specific_context, context_dir
+                )
+                annotation_list.append(_json_o)
+                specific_context.update(_local_context)
+            json_object[k] = annotation_list
+        # Add the key mapping to the context
+        if k not in context.keys():
+            logger.debug(f'{k} not in context: {pal.Path(schema["id"]).name}')
+            continue
+        if k not in specific_context.keys():
+            specific_context[k] = context[k]
+            logger.debug(f"{k} added to context: {context[k]}")
+        elif specific_context[k] == context[k]:
+            continue
+        else:
+            logger.debug(f"{k} duplicate: {specific_context[k]} vs {context[k]}")
+            continue
+
+    dtype = schema["properties"]["@type"]["enum"][0]
+    if dtype not in context.keys():
+        logger.debug(f'{dtype} not in context {schema["id"]}')
+    else:
+        specific_context[dtype] = context[dtype]
+    json_object["@type"] = dtype
+    return json_object, specific_context
+
+
+def gen_jsonld_outpath(dats_json_f, out_path):
+    """
+    This function generates a output file path for the annotated DATS jsonld file.
+
+    :param dats_json_f: the path to the original DATS instance file
+    :param out_path: the folder or new file path where the annotated DATS.jsonld file should be stored
+    :return:  the final DATS.jsonld file path
+    """
+    if not isinstance(dats_json_f, pal.Path):
+        dats_json_f = pal.Path(dats_json_f)
+    if out_path is not None and not isinstance(out_path, pal.Path):
+        out_path = pal.Path(out_path)
+
+    out_name = dats_json_f.with_suffix(".jsonld").name
+    if out_path is None:
+        # We save the output to the parent directory of the current DATS JSON file
+        out_path = dats_json_f.parent
+    elif out_path.suffix != "":
+        # We have most likely gotten a path to a non-existent file, let's use that
+        return out_path
+    elif not out_path == dats_json_f.parent:
+        # We are saving this somewhere other than the original folder
+        out_name = f"{dats_json_f.parent.name}_{out_name}"
+    if out_path.is_dir():
+        dats_jsonld_f = out_path / out_name
+    elif out_path.parent.is_dir():
+        # TODO: check if there is a way to explicitly look for a file that doesn't exist
```

Comment:
wouldn't `my_file.is_file()` or `my_file.exists()` do this?

Reply:
I don't think so but I also think that this comment is outdated. My goal was to find some check whether the submitted path points to a file. The solution I came up with is to check for a non-empty, dot-separated file suffix ([in line 189](https://github.com/CONP-PCNO/conp-dataset/pull/640/files#diff-e5ad809a7eb9c3c0fa56ab26cd60aab38eb52f7af96b096b0302464988592650R189)).  

I think this section is not doing a good job. The only way to arrive here is if you supplied the path to a non-existent directory, e.g. `/all/these/dirs/exist/but_not_this_one`. Then the current code would put the results in `/all/these/dirs/exist/` which is most likely unexpected. So maybe here we should rather create the directory `but_not_this_one`. What do you think @cmadjar